### ISSUE TITLE
Getting masks from SAM3 model

### ISSUE THREAD

**Issue by LightCannon** [jamjamjon/usls#174]
Hello.
1- Is there any example to get the masks out of the SAM3 model? the existing example annotates only bounding boxes.
2- Also, the saved annotated image is always larger than the actual input image.

**Comment by jamjamjon** [comment]
@LightCannon Hi there!

For the first question:

### **1. “Is there any example to get the masks out of the SAM3 model? The existing example only annotates bounding boxes.”**

You can enable mask visualization by adding `--show-mask`:

```bash
cargo run -F sam3,cuda --example sam3 -- --device cuda -p shoe --show-mask
```

You will see output like this:

![Image](https://github.com/user-attachments/assets/0acbf1bc-f5b5-4672-90b3-851720edced4)

The masks are available in `Ys` → `Y` as `image::GrayImage`.

You can extract the raw mask data using `mask.to_vec()`
and obtain the contour using `mask.polygon()`:

```rust
// Process in batches
for batch in dataloader {
    let ys = model.forward(&batch, &prompts)?;

    for (img, y) in batch.iter().zip(ys.iter()) {
        for (i, mask) in y.masks.iter().enumerate() {
            let mask_vec = mask.to_vec();
            let mask_contour = mask.polygon();
        }
    }
}
```

---

### **2. “Does the mask size match the original image size?”**

Yes. The mask dimensions always match the original image:

```rust
for batch in dataloader {
    let ys = model.forward(&batch, &prompts)?;

    for (img, y) in batch.iter().zip(ys.iter()) {
        for (i, mask) in y.masks.iter().enumerate() {
            println!("{} => original image size: {:?}, mask {:?}", 
                     i, img.dimensions(), mask);
        }
    }
}
```

Example output:

```
0 => original image size: (1280, 720), mask Mask { dimensions: (1280, 720), id: 0, name: "shoe", confidence: 0.91564703 }
1 => original image size: (1280, 720), mask Mask { dimensions: (1280, 720), id: 0, name: "shoe", confidence: 0.8862287 }
...
```

The image and mask resolutions are exactly the same.

**Comment by LightCannon** [comment]
@jamjamjon 
Then why I'm getting only black images for the masks? you can see above that I'm getting NaNs when I checked the impl. internally.

Currently, whatever image I pass, I could not get a mask, just boxes. The rest of images are black images (so masks are just black images)

**Comment by jamjamjon** [comment]
@LightCannon I see what’s going on — you’re looking at the wrong mask.
The NaNs you saw come from query 0, but that doesn’t mean the model or ONNX export is broken.

## Why mask[0] is NaN

The SAM3 decoder outputs 200 object queries:
```
pred_logits: [1,200]
pred_boxes:  [1,200,4]
pred_masks:  [1,200,H,W]
```
Just like DETR, most of these queries are unused.
Query 0 is usually background or unactivated, so its mask can be NaN, all zeros, or random noise — this is completely normal.

## How the Rust code handles masks

We never use mask[0] directly. Masks are selected based on their scores:
```rust
let presence_score = 1.0 / (1.0 + (-presence_logits[[0,0]]).exp());
let scores = pred_logits.slice(s![0,..])
    .mapv(|x| 1.0 / (1.0 + (-x).exp()) * presence_score);

let valid_indices: Vec<usize> = scores.iter()
    .enumerate()
    .filter(|(_, &s)| s >= self.conf[0])
    .map(|(i, _)| i)
    .collect();
```
Only the queries with high scores are used; low-score queries (including index 0) are ignored.

## Why you saw black masks

Because you manually inspected:`slice(s![0, 0, ...])`. That query isn’t a real object.
If you check the mask corresponding to the highest-score query, you’ll see a valid mask.

**Comment by LightCannon** [comment]
What got me to check is that I found the mask images black. 
So, with the last code committed, I'm still getting black masks.

**Comment by LightCannon** [comment]
@jamjamjon 
Here is what I'm doing

this is the main.rs (exactly like the example, just added more prints)
`use anyhow::Result;
use usls::{
    models::{Sam3Prompt, SAM3},
    Annotator, Config, DataLoader,
};

#[derive(argh::FromArgs)]
/// SAM3 - Segment Anything Model 3
struct Args {
    /// device (cpu:0, cuda:0, etc.)
    #[argh(option, default = "String::from(\"cpu:0\")")]
    device: String,

    /// source image paths (can specify multiple)
    #[argh(option, default = "vec![String::from(\"./assets/sam3-demo.jpg\")]")]
    source: Vec<String>,

    /// prompts: "text;pos:x,y,w,h;neg:x,y,w,h" (can specify multiple)
    #[argh(option, short = 'p')]
    prompt: Vec<String>,

    /// confidence threshold (default: 0.5)
    #[argh(option, default = "0.5")]
    conf: f32,

    /// show mask
    #[argh(switch)]
    show_mask: bool,

    /// dtype
    #[argh(option, default = "String::from(\"q4f16\")")]
    dtype: String,

    /// vision encoder batch min
    #[argh(option, default = "1")]
    vision_batch_min: usize,

    /// vision encoder batch opt
    #[argh(option, default = "1")]
    vision_batch: usize,

    /// vision encoder batch max
    #[argh(option, default = "8")]
    vision_batch_max: usize,

    /// text encoder batch min
    #[argh(option, default = "1")]
    text_batch_min: usize,

    /// text encoder batch opt
    #[argh(option, default = "4")]
    text_batch: usize,

    /// text encoder batch max
    #[argh(option, default = "16")]
    text_batch_max: usize,

    /// geometry encoder batch min
    #[argh(option, default = "1")]
    geo_batch_min: usize,

    /// geometry encoder batch opt
    #[argh(option, default = "8")]
    geo_batch: usize,

    /// geometry encoder batch max
    #[argh(option, default = "16")]
    geo_batch_max: usize,

    /// decoder batch min
    #[argh(option, default = "1")]
    decoder_batch_min: usize,

    /// decoder batch opt
    #[argh(option, default = "1")]
    decoder_batch: usize,

    /// decoder batch max
    #[argh(option, default = "8")]
    decoder_batch_max: usize,
}

fn main() -> Result<()> {
    tracing_subscriber::fmt()
        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
        .with_timer(tracing_subscriber::fmt::time::ChronoLocal::rfc_3339())
        .init();
    let args: Args = argh::from_env();

    // Parse prompts
    if args.prompt.is_empty() {
        anyhow::bail!("No prompt. Use -p \"text\" or -p \"visual;pos:x,y,w,h\"");
    }
    let prompts: Vec<Sam3Prompt> = args
        .prompt
        .iter()
        .map(|s| s.parse())
        .collect::<std::result::Result<Vec<_>, _>>()
        .map_err(|e| anyhow::anyhow!("{}", e))?;

    // Build model with per-encoder batch sizes
    let config = Config::sam3_image_predictor()
        .with_dtype_all(args.dtype.parse()?)
        .with_class_confs(&[args.conf])
        // Per-encoder batch sizes for TensorRT (min, opt, max)
        .with_visual_encoder_batch_min_opt_max(
            args.vision_batch_min,
            args.vision_batch,
            args.vision_batch_max,
        )
        .with_textual_encoder_batch_min_opt_max(
            args.text_batch_min,
            args.text_batch,
            args.text_batch_max,
        )
        .with_encoder_batch_min_opt_max(args.geo_batch_min, args.geo_batch, args.geo_batch_max)
        .with_decoder_batch_min_opt_max(
            args.decoder_batch_min,
            args.decoder_batch,
            args.decoder_batch_max,
        )
        // Device configuration
        // => If your GPU memory is insufficient, you can place some modules on CPU
        // .with_visual_encoder_device(args.device.parse()?)
        // .with_textual_encoder_device(args.device.parse()?)
        // .with_encoder_device(args.device.parse()?)  // geometry-encoder
        // .with_decoder_device(args.device.parse()?)
        .with_device_all(args.device.parse()?)
        .commit()?;
    let mut model = SAM3::new(config)?;

    // Annotator
    let annotator = Annotator::default().with_mask_style(
        usls::Style::mask()
            .with_draw_mask_polygon_largest(true)
            .with_visible(args.show_mask),
    );
    let output_dir = usls::Dir::Current.base_dir_with_subs(&["runs", model.spec()])?;

    // DataLoader with batch iteration
    let dataloader = DataLoader::from_paths(&args.source)?
        .with_batch(args.vision_batch)
        .with_progress_bar(true)
        .build()?;

    // Process in batches
    for batch in dataloader {
        let ys = model.forward(&batch, &prompts)?;
        println!("ys: {:?}", ys);

        for (img, y) in batch.iter().zip(ys.iter()) {
            annotator
                .annotate(img, y)?
                .save(output_dir.join(format!("{}.jpg", usls::timestamp(None))))?;
        }
    }

    // Performance stats
    usls::perf(false);

    Ok(())
}`


cargo.toml
`[package]
name = "sam3_test"
version = "0.1.0"
edition = "2024"

[dependencies]
anyhow = "1.0.100"
argh = "0.1.13"
tracing-subscriber = { version = "0.3.18", features = ["env-filter", "chrono"] }
usls = { git = "https://github.com/jamjamjon/usls", features = [ "cuda", "sam3"] }


[target.x86_64-pc-windows-msvc]
rustflags = ["-C", "target-feature=-crt-static"]

[patch.crates-io]
esaxx-rs = { path = "./patches/esaxx-rs" }`

The command: cargo run --  --device cuda --dtype q4f16 --source "C:\Users\MMD\Downloads\test_image.jpg" -p "shoe" --show-mask
You can see the result: 
- Bounding boxes are there
- No masks, their corresponding images are black images
![Image](https://github.com/user-attachments/assets/1c38e4c6-9129-4df2-a678-8e1472cf1673)
